File: plugins/tokenmonitor/server/js/test/textutil.test.js
import { test } from "node:test";
import assert from "node:assert/strict";

import { clipCodePoints } from "../src/textutil.js";

test("clipCodePoints counts code points, not UTF-16 units", () => {
  assert.equal(clipCodePoints("", 15), "");
  assert.equal(clipCodePoints("short", 15), "short");
  assert.equal(clipCodePoints("exactly15chars!", 15), "exactly15chars!");
  assert.equal(clipCodePoints("sixteen chars!!!", 15), "sixteen chars!!");
  // 15 code points with a multibyte char — must survive intact.
  assert.equal(clipCodePoints("Dragón de fuego", 15), "Dragón de fuego");
  assert.equal(clipCodePoints("Dragón de fuegoX", 15), "Dragón de fuego");
  // Astral (surrogate-pair) code points count as one each; a UTF-16
  // .slice(0, 2) here would split the second dragon into a lone surrogate.
  assert.equal(clipCodePoints("🐉🐉🐉", 2), "🐉🐉");
  assert.equal(clipCodePoints("🐉🐉🐉", 15), "🐉🐉🐉");
  assert.equal(clipCodePoints("abc", 0), "");
  // Never emits a lone surrogate.
  for (const cut of ["🐉🐉🐉", "a🐉b🐉", "🐉aaaaaaaaaaaaaa🐉"]) {
    const out = clipCodePoints(cut, 15);
    assert.ok(!/[\uD800-\uDBFF]$/.test(out), `lone surrogate at end of ${JSON.stringify(out)}`);
  }
});

File: plugins/tokenmonitor/server/go/internal/textutil/textutil_test.go
package textutil

import (
	"testing"
	"unicode/utf8"
)

func TestClipRunes(t *testing.T) {
	cases := []struct {
		in   string
		n    int
		want string
	}{
		{"", 15, ""},
		{"short", 15, "short"},
		{"exactly15chars!", 15, "exactly15chars!"},
		{"sixteen chars!!!", 15, "sixteen chars!!"},
		// Multibyte: "Dragón de fuego" is 15 code points, 16 bytes.
		{"Dragón de fuego", 15, "Dragón de fuego"},
		{"Dragón de fuegoX", 15, "Dragón de fuego"},
		// Byte-based s[:15] would split the ó (2 bytes at index 14).
		{"aaaaaaaaaaaaaaó", 15, "aaaaaaaaaaaaaaó"},
		{"aaaaaaaaaaaaaaóZ", 15, "aaaaaaaaaaaaaaó"},
		// Astral (4-byte) code points count as one each.
		{"🐉🐉🐉", 2, "🐉🐉"},
		{"🐉🐉🐉", 15, "🐉🐉🐉"},
		{"abc", 0, ""},
	}
	for _, c := range cases {
		got := ClipRunes(c.in, c.n)
		if got != c.want {
			t.Errorf("ClipRunes(%q, %d) = %q, want %q", c.in, c.n, got, c.want)
		}
		if !utf8.ValidString(got) {
			t.Errorf("ClipRunes(%q, %d) produced invalid UTF-8: %q", c.in, c.n, got)
		}
	}
}
